src/network/connection.js

```diff
@@ -304,43 +304,43 @@ module.exports = class Connection {
 
     this.buffer = Buffer.concat([this.buffer, rawData])
 
-    // Not enough bytes to read the expected response size, keep buffering
-    if (Buffer.byteLength(this.buffer) <= Decoder.int32Size()) {
-      return
-    }
-
-    const data = Buffer.from(this.buffer)
-    const decoder = new Decoder(data)
-    const expectedResponseSize = decoder.readInt32()
+    // Process data if there are enough bytes to read the expected response size,
+    // otherwise keep buffering
+    while (Buffer.byteLength(this.buffer) > Decoder.int32Size()) {
+      const data = Buffer.from(this.buffer)
+      const decoder = new Decoder(data)
+      const expectedResponseSize = decoder.readInt32()
+
+      if (!decoder.canReadBytes(expectedResponseSize)) {
+        return
+      }
 
-    if (!decoder.canReadBytes(expectedResponseSize)) {
-      return
-    }
+      const response = new Decoder(decoder.readBytes(expectedResponseSize))
+      // Reset the buffer as the rest of the bytes
+      this.buffer = decoder.readAll()
 
-    // The full payload is loaded, erase the temporary buffer
-    this.buffer = Buffer.alloc(0)
+      if (this.authHandlers) {
+        return this.authHandlers.onSuccess(data.slice(0, this.offset + byteLength))
+      }
 
-    if (this.authHandlers) {
-      return this.authHandlers.onSuccess(data)
-    }
+      const correlationId = response.readInt32()
+      const payload = response.readAll()
 
-    const correlationId = decoder.readInt32()
-    const payload = decoder.readAll()
+      const entry = this.pendingQueue[correlationId]
+      delete this.pendingQueue[correlationId]
```

Comment:
I wonder if we should remove the previous data from the buffer, I fear to end up leaking memory.

Reply:
I think we are okey, since the buffer is reset using `decoder.readAll()` which only copies (not slicing) the rest of data. So the original buffer could be collected by GC once the response gets consumed.